Fixed ref conf calculation near indels

ldgauthier opened the pull request:
Previous behavior generated some PL=0,0,0 no-calls because CIGAR of reads containing indels wasn't taken into account when determining which reads were informative for the indel ref conf model.  The local realignment wasn't being used inside the active region previously either, which has been fixed. A related change considers bases on either side of indels informative if local assembly has been performed (but not during active region detection).  Both result in far fewer 0,0,0 calls.  Unfortunately there are still some 0,0,X homRef calls related to #5171.

ldgauthier commented:
@droazen The HCSpark test failures are admittedly my fault, but there's a GDB test that's failing on Travis but passing locally, just like in my fixMQcalc PR.  I'd like both of these to go into a release soon so we can set the GATK version for the exomes in the cloud pipeline.

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java:
```diff
@@ -196,6 +196,56 @@ public static GATKRead createReadAlignedToRef(final GATKRead originalRead,
+        return getSequenceAlignedOneToOne(read, false, (byte)0);
+    }
+
+    public static byte[] getSequenceAlignedOneToOne(final GATKRead read, final boolean forBases, final byte padWith) {
```
Using a boolean param to indicate whether this is an operation on quality scores or read bases smells a bit. What about provide a lambda to obtain the original-not-copied byte array given the read. 
```
public static byte[] getSequenceAlignedOneToOne(final GATKRead read, Function<GATKRead, byte[]> bytesProvider, int padWith) {
   ...
} 
```

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java:
```diff
@@ -196,6 +196,56 @@ public static GATKRead createReadAlignedToRef(final GATKRead originalRead,
+            int paddedPos = 0;
+            for ( int i = 0; i < cigar.numCigarElements(); i++ ) {
+                final CigarElement ce = cigar.getCigarElement(i);
+                switch ( ce.getOperator() ) {
```
This switch can be rewriten with a couple of ifs and using the ```consumesReadBases``` and ```consumesReferenceBases``` without the need for list operators explicitly.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/ReferenceConfidenceModel.java:
```diff
@@ -547,7 +558,7 @@ int sumMismatchingQualities(final byte[] readBases,
             final byte readBase = readBases[readStart + i];
             final byte refBase  = refBases[refStart + i];
-            if ( readBase != refBase ) {
+            if ( readBase != refBase && readBase != 'N') {
```
These are not totally equivalent but I guess it should work as intended (e.g. the Nucleotide. alternative would handle other ambiguity codes such as P R, W etc.)

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/ReferenceConfidenceModel.java:
```diff
@@ -635,6 +645,38 @@ int calcNIndelInformativeReads(final ReadPileup pileup, final int pileupOffsetIn
+     * @return the new reference-aligned index/offset
+     */
+    @VisibleForTesting
+    protected int getCigarModifiedOffset (final PileupElement p){
```
I didn't use those cigar methods because matches consume read and reference bases, which is not what I want.

vruano commented:
Looking back into this PR... at some point you are using 'N' to pad what seem to be gaps on the read sequence. Although the end result would be the same perhaps is better to be more explicit and just use '-' instead. In that case my suggestion of using `Nucleotide.intersect` wouldn't cover for the  '-' character so you need a explicity "&&" or "II".

When you compare the cost of each different alignment the gap-open and gap-ext are ignored (you only look at base call mismatches). I wonder whether it would be more correct to actually take them in consideration... so imagine that there is no gaps in the original alignment what-soever and that adding a 1bp gap decreases the number of mis-matches by just 1 which is typically Q30 increase in the Lk but the gap itself default penalty is Q45 so can one say that that read wouldn't still support the reference over a 1-bp gap alternative? 

Example with a 2-bp gap making the trick:
```
Ref:  ....GCATGTGATATATATATATATATATATATACACACAC....
Read: ....GCATGTG--ATATATATATATATATATATAC <end-of-the-read>
```

That could happen in STRs with impurities... but if the original alignment did not added itself the gap to reduce the number of mismatches is because precisely due to the added cost of the gap-open and necessary extends that we would be ignoring here.

This is all hypothetical until some one quantifies how often this might occur ... so I'm happy to keep ignoring the gaps for now until we get a report on a real-live dataset that would benefit of such a change or some enthusiastic dsde-methods member investigates this.

ldgauthier commented:
@vruano I addressed your comments.  The code is a lot cleaner.  Let me know if there's anything else.  The Travis failures look unrelated so I'm rerunning them.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/ReferenceConfidenceModel.java:
```diff
@@ -555,7 +555,8 @@ int sumMismatchingQualities(final byte[] readBases,
             final byte readBase = readBases[readStart + i];
             final byte refBase  = refBases[refStart + i];
-            if ( !Nucleotide.decode(readBase).intersect(Nucleotide.decode(refBase)).isValid()) {  //allow N read bases (and other ambiguous bases) to "match" reference
+            if ( !Nucleotide.decode(readBase).intersect(Nucleotide.decode(refBase)).isValid() &&
```
Ah I see, I was assume that master had a change that I have on Nucleotides adding an new intersect(byte, byte) method:
```
/**
 * Checks whether two nucleotides intersect given their byte encodings.
 * @param a first nucleotide.
 * @param b second nucleotide.
 * @return {@code true} iff the input nucleotides intersect.
 */
public static boolean intersect(final byte a, final byte b) {
  return (baseToValue[0xFF & a].mask & baseToValue[0xFF & b].mask) != 0;
}
```

Just add it in your branch's Nucleotides.java and the the if code would be:

```
if (!Nucleotide.intersect(readBase, refBase) && readBase != AlignmentUtils.PLACEHOLDER) {
   ....
}
```

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java:
```diff
@@ -21,6 +21,7 @@
     private static final EnumSet<CigarOperator> ALIGNED_TO_GENOME_OPERATORS = EnumSet.of(CigarOperator.M, CigarOperator.EQ, CigarOperator.X);
     private static final EnumSet<CigarOperator> ALIGNED_TO_GENOME_PLUS_SOFTCLIPS = EnumSet.of(CigarOperator.M, CigarOperator.EQ, CigarOperator.X, CigarOperator.S);
     public final static String HAPLOTYPE_TAG = "HC";
+    public final static char PLACEHOLDER = '-';
```
I would call it GAP or GAP_CHARACTER.

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java:
```diff
@@ -199,7 +200,7 @@ public static GATKRead createReadAlignedToRef(final GATKRead originalRead,
 
     public static byte[] getBasesAlignedOneToOne(final GATKRead read) {
-        return getSequenceAlignedOneToOne(read, r -> r.getBasesNoCopy(), Nucleotide.N.encodeAsByte());
+        return getSequenceAlignedOneToOne(read, r -> r.getBasesNoCopy(), (byte)PLACEHOLDER);
```
Perhaps GAP or PLACEHOLDER should be `byte` typed rather than `char`.

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java:
```diff
@@ -198,25 +199,21 @@ public static GATKRead createReadAlignedToRef(final GATKRead originalRead,
 
-    public static byte[] getSequenceAlignedOneToOne(final GATKRead read, final boolean forBases, final byte padWith) {
+    public static byte[] getSequenceAlignedOneToOne(final GATKRead read, final Function<GATKRead, byte[]> bytesProvider, final byte padWith) {
         final Cigar cigar = read.getCigar();
```
Do null checking on read and baseProvider.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/ReferenceConfidenceModel.java:
```diff
@@ -654,25 +651,12 @@ int calcNIndelInformativeReads(final ReadPileup pileup, final int pileupOffsetIn
-            if (cigar.getCigarElement(i).getOperator().equals(CigarOperator.INSERTION)) {
-                offset -= cigar.getCigarElement(i).getLength();
+        int offset = p.getCurrentCigarElement().getOperator().consumesReferenceBases() ? p.getOffsetInCurrentCigar() : 0;
+        for (CigarElement elem : read.getCigar().getCigarElements().subList(0, p.getCurrentCigarOffset())) {
```
Sorry, I didn't pay close attention that my suggested code was totally correct... the ```int offset = ``` condition ... do you need to add ```|| elem.getOperator() == CigarOperator.S)```?

vruano replied:
Make ```elem``` ```final```.

vruano commented on src/main/java/org/broadinstitute/hellbender/utils/read/CigarUtils.java:
```diff
@@ -106,7 +106,7 @@ public static int countRefBasesBasedOnCigar(final GATKRead read, final int cigar
             final CigarOperator operator = cigarElement.getOperator();
-            if (operator.consumesReferenceBases() || operator == CigarOperator.S) {
+            if (operator.consumesReferenceBases() || operator.isClipping()) {
                 result += cigarElement.getLength();
```
I see that given the original code, this is the wright translation so no changes needed... but the documentation is rather unclear about what this method does. Perhaps a better name for it would be ```countReferenceBasesOnUnclippedAlignment```.

My point here is that now we have a some methods that count soft and hard-clips in this kind of operations whereas some other methods only du soft... and I wonder whether there is a need for consolidate these (e.g. perhaps some changes in so calling code would make such a difference unnecessary). 

In any case feel free to leave it for a future PR.

vruano commented:
@ldgauthier I asked for further changes in my latest review (can you see it). By accident I approved rather than ask for changes so please take a look before merging. It should not take long to address them. Then you can merge.
